service/history/replication/raw_task_converter.go

```diff
@@ -712,3 +637,323 @@ func convertGetHistoryError(
+	lca *historyspb.VersionHistoryItem,
+) (*replicationspb.ReplicationTask, error) {
+	events, next_page_token, err := c.getEvents(
+		transitionHistory,
+		lca.GetEventId(),
+		lca.GetVersion(),
+		false,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return &replicationspb.ReplicationTask{
+		TaskType:     enumsspb.REPLICATION_TASK_TYPE_SYNC_VERSIONED_TRANSITION_TASK,
+		SourceTaskId: c.taskInfo.TaskID,
+		Priority:     c.taskInfo.Priority,
+		Attributes: &replicationspb.ReplicationTask_SyncVersionedTransitionTaskAttributes{
+			SyncVersionedTransitionTaskAttributes: &replicationspb.SyncVersionedTransitionTaskAttributes{
+				EventBatches:  events,
+				NextPageToken: next_page_token,
+				// TODO: SyncStateSnapshotAttributes
+			},
+		},
+		VersionedTransition: transitionHistory[len(transitionHistory)-1],
+		VisibilityTime:      timestamppb.New(c.taskInfo.VisibilityTimestamp),
+	}, nil
+}
+
+func (c *syncVersionedTransitionTaskConverter) generateVerifyVersionedTransitionTask() (*replicationspb.ReplicationTask, error) {
+	versionHistoryIndex, err := versionhistory.FindFirstVersionHistoryIndexByVersionHistoryItem(
+		c.mutableState.GetExecutionInfo().VersionHistories,
+		versionhistory.NewVersionHistoryItem(
+			c.taskInfo.FirstEventID,
+			c.taskInfo.VersionedTransition.NamespaceFailoverVersion,
+		),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return &replicationspb.ReplicationTask{
+		TaskType:     enumsspb.REPLICATION_TASK_TYPE_VERIFY_VERSIONED_TRANSITION_TASK,
+		SourceTaskId: c.taskInfo.TaskID,
+		Attributes: &replicationspb.ReplicationTask_VerifyVersionedTransitionTaskAttributes{
+			VerifyVersionedTransitionTaskAttributes: &replicationspb.VerifyVersionedTransitionTaskAttributes{
+				NamespaceId:         c.taskInfo.NamespaceID,
+				WorkflowId:          c.taskInfo.WorkflowID,
+				RunId:               c.taskInfo.RunID,
+				NewRunId:            c.taskInfo.NewRunID,
+				EventVersionHistory: c.mutableState.GetExecutionInfo().VersionHistories.Histories[versionHistoryIndex].Items,
+				NextEventId:         c.taskInfo.NextEventID,
+			},
+		},
+		VersionedTransition: c.taskInfo.VersionedTransition,
+		VisibilityTime:      timestamppb.New(c.taskInfo.VisibilityTimestamp),
+	}, nil
+}
+
+func (c *syncVersionedTransitionTaskConverter) generateBackfillHistoryTask() (*replicationspb.ReplicationTask, error) {
+	taskVersionHistoryItems, taskEvents, taskNewEvents, _, err := getVersionHistoryAndEventsWithNewRun(
+		c.ctx,
+		c.shardContext,
+		c.shardID,
+		definition.NewWorkflowKey(c.taskInfo.NamespaceID, c.taskInfo.WorkflowID, c.taskInfo.RunID),
+		c.taskInfo.VersionedTransition.NamespaceFailoverVersion,
+		c.taskInfo.FirstEventID,
+		c.taskInfo.NextEventID,
+		c.taskInfo.NewRunID,
+		c.workflowCache,
+		c.eventBlobCache,
+		c.executionManager,
+		c.logger,
+	)
+	if err != nil {
+		return nil, err
+	}
+	err = c.replicationCache.Update(c.taskInfo.RunID, c.targetClusterID, nil, taskVersionHistoryItems)
```

Comment:
is `taskVersionHistoryItems` only up to the nextEventID this task is trying to backfill?

Reply:
Good point. taskVersionHistoryItems will be the current versionHistoryItems containing the first event and version. Need to truncate to the task.nextEventID

Reply:
updated.